File: backend/src/main/java/com/back/global/rq/Rq.java
package com.back.global.rq;

import com.back.domain.member.member.entity.Member;
import com.back.global.security.SecurityUser;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;

import java.util.Optional;

@Component
@RequiredArgsConstructor
public class Rq {
    private final HttpServletRequest req;
    private final HttpServletResponse resp;

    public Member getActor() {
        return Optional.ofNullable(SecurityContextHolder.getContext().getAuthentication())
                .map(Authentication::getPrincipal)
                .filter(SecurityUser.class::isInstance)
                .map(SecurityUser.class::cast)
                .map(SecurityUser::getMember)
                .orElse(null);
    }

    public void setCookie(String name, String value) {
        if (value == null) value = "";

        String domain = req.getServerName().contains("localhost") ? "localhost" : "bookers.p-e.kr";

        Cookie cookie = new Cookie(name, value);
        cookie.setPath("/");
        cookie.setHttpOnly(true);
        cookie.setSecure(true);
        cookie.setAttribute("SameSite", "None");
        cookie.setDomain(domain);

        if(value.isEmpty()){
            cookie.setMaxAge(0); // 즉시 만료
        } else {
            if(name.equals("accessToken")){
                cookie.setMaxAge(60*20); //20분
            } else if (name.equals("refreshToken")) {
                cookie.setMaxAge(60 * 60 * 24); // 1일
            } else {
                cookie.setMaxAge(60 * 60 * 24 * 365); // 1년
            }
        }
        resp.addCookie(cookie);
    }

    public void clearAuthCookies() {
        setCookie("accessToken", "");
        setCookie("refreshToken", "");
    }

}

File: backend/src/main/java/com/back/global/rq/Rq.kt
package com.back.global.rq

import com.back.domain.member.member.entity.Member
import com.back.domain.member.member.service.MemberService
import com.back.global.security.SecurityUser
import com.back.standard.extensions.getOrThrow
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component


@Component
class Rq(
    private val req: HttpServletRequest,
    private val resp: HttpServletResponse,
    private val memberService: MemberService
) {
    companion object {
        private const val ACCESS_TOKEN_DURATION = 60 * 20 // 20분
        private const val REFRESH_TOKEN_DURATION = 60 * 60 * 24 // 1일
        private const val DEFAULT_TOKEN_DURATION = 60 * 60 * 24 * 365 // 1년

        private const val LOCALHOST = "localhost"
        private const val PRODUCTION_DOMAIN = "bookers.p-e.kr"
    }

    val actor: Member?
        get() = SecurityContextHolder.getContext().authentication
            ?.principal
            ?.let { it as? SecurityUser }
            ?.member

    val actorFromDb: Member
            get() = memberService.findById(getAuthenticatedActor().id).getOrThrow()

    fun getAuthenticatedActor(): Member = actor
            ?: throw IllegalStateException("인증된 사용자 정보를 찾을 수 없습니다")

    fun setCookie(name: String, value: String?) {
        val cookieValue = value ?: ""

        val domain = if (req.serverName.contains(LOCALHOST)) LOCALHOST else PRODUCTION_DOMAIN

        val cookie = Cookie(name, cookieValue).apply{
            path = "/"
            isHttpOnly = true
            secure = true
            setAttribute("SameSite", "None")
            setDomain(domain)
            maxAge = when {
                cookieValue.isEmpty() -> 0 // 즉시 만료
                name == "accessToken" -> ACCESS_TOKEN_DURATION
                name == "refreshToken" -> REFRESH_TOKEN_DURATION
                else -> DEFAULT_TOKEN_DURATION
            }
        }
        resp.addCookie(cookie)
    }

    fun clearAuthCookies() {
        listOf("accessToken", "refreshToken").forEach { name ->
            setCookie(name, "")
        }
    }

    fun sendRedirect(url: String) {
        resp.sendRedirect(url)
    }
}
